src/schema/create-index-builder.ts

```diff
@@ -126,6 +133,101 @@ export class CreateIndexBuilder implements OperationNodeSource, Compilable {
     })
   }
 
+  /**
+   * Adds a where clause to the query. This Effectively turns the index partial.
+   *
+   * Also see {@link orWhere}
+   *
+   * ### Examples
+   *
+   * ```ts
+   * import { sql } from 'kysely'
+   *
+   * await db.schema
+   *    .createIndex('orders_unbilled_index')
+   *    .on('orders')
+   *    .column('order_nr')
+   *    .where(sql.ref('billed'), 'is not', true)
+   *    .where('order_nr', 'like', '123%')
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * create index "orders_unbilled_index" on "orders" ("order_nr") where "billed" is not true and "order_nr" like '123%'
+   * ```
+   *
+   * Column names specified in {@link column} or {@link columns} are known at compile-time
+   * and can be referred to in the current query and context.
+   *
+   * Sometimes you may want to refer to columns that exist in the table but are not
+   * part of the current index. In that case you can refer to them using {@link sql}
+   * expressions.
+   *
+   * Parameters are always sent as literals due to database restrictions.
+   */
+  where(
+    lhs: C | Expression<any>,
+    op: ComparisonOperatorExpression,
+    rhs: unknown
+  ): CreateIndexBuilder<C>
+  where(
+    grouper: (qb: CreateIndexWhereInterface<C>) => CreateIndexWhereInterface<C>
+  ): CreateIndexBuilder<C>
+  where(expression: Expression<any>): CreateIndexBuilder<C>
+
+  where(...args: any[]): any {
+    return new CreateIndexBuilder({
+      ...this.#props,
+      node: QueryNode.cloneWithWhere(
+        this.#props.node as any,
```

Comment:
Instead of casting to `any` change the `cloneWithWhere` interface from:

```ts
  cloneWithWhere<T extends FilterableQueryNode>(
    node: T,
    operation: OperationNode
  ): T
```

to something like this:

```ts
  cloneWithWhere<T extends { where?: WhereNode }>(
    node: T,
    operation: OperationNode
  ): T
```

Reply:
Good idea!